drivers/sensor/tsl2591/tsl2591.c

```diff
@@ -0,0 +1,534 @@
+		return 0;
+	}
+
+	return tsl2591_reg_write(dev, reg, new_value);
+}
+
+static int tsl2591_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct tsl2591_data *data = dev->data;
+	uint8_t als_data[4];
+	int ret;
+
+#ifdef CONFIG_TSL2591_FETCH_WAIT
+	uint8_t status;
+
+	ret = tsl2591_reg_read(dev, TSL2591_REG_STATUS, &status, 1U);
+	if (ret < 0) {
+		LOG_ERR("Failed to read status register");
+		return ret;
+	}
+
+	/* Check if ALS has completed an integration cycle since AEN asserted.
+	 * If not, sleep for the duration of an integration cycle to ensure valid reading.
+	 */
+	if ((status & TSL2591_AVALID_MASK) != 1) {
+		k_msleep((data->atime / 100) * TSL2591_MAX_TIME_STEP);
+	}
+
+	/* Reassert AEN to determine if next reading is valid */
+	ret = tsl2591_reg_update(dev, TSL2591_REG_ENABLE, TSL2591_AEN_MASK, TSL2591_AEN_OFF);
+	if (ret < 0) {
+		LOG_ERR("Failed to disable ALS");
+		return ret;
+	}
+
+	ret = tsl2591_reg_update(dev, TSL2591_REG_ENABLE, TSL2591_AEN_MASK, TSL2591_AEN_ON);
+	if (ret < 0) {
+		LOG_ERR("Failed to re-enable ALS");
+		return ret;
+	}
+#endif
+
+	switch (chan) {
+	case SENSOR_CHAN_ALL:
+		ret = tsl2591_reg_read(dev, TSL2591_REG_C0DATAL, als_data, 4U);
+		if (ret < 0) {
+			LOG_ERR("Failed to read ALS data");
+			return ret;
+		}
+
+		data->vis_count = sys_get_le16(als_data);
+		data->ir_count = sys_get_le16(als_data + 2);
+		break;
+	case SENSOR_CHAN_LIGHT:
+		ret = tsl2591_reg_read(dev, TSL2591_REG_C0DATAL, als_data, 2U);
+		if (ret < 0) {
+			LOG_ERR("Failed to read ALS visible light data");
+			return ret;
+		}
+
+		data->vis_count = sys_get_le16(als_data);
+		break;
+	case SENSOR_CHAN_IR:
+		ret = tsl2591_reg_read(dev, TSL2591_REG_C1DATAL, als_data, 2U);
+		if (ret < 0) {
+			LOG_ERR("Failed to read ALS infrared data");
+			return ret;
+		}
+
+		data->ir_count = sys_get_le16(als_data);
+		break;
+	default:
+		LOG_ERR("Unsupported sensor channel");
+		return -ENOTSUP;
+	}
+
+#ifdef CONFIG_TSL2591_WARN_SATURATED
+	uint16_t max_count = data->atime == 100 ? TSL2591_MAX_ADC_100 : TSL2591_MAX_ADC;
+	bool vis_saturated = (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_LIGHT) &&
+			     (data->vis_count >= max_count);
+	bool ir_saturated = (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_IR) &&
+			    (data->ir_count >= max_count);
+	if (vis_saturated || ir_saturated) {
+		LOG_WRN("Sensor ADC potentially saturated, reading may be invalid");
+		return -EOVERFLOW;
+	}
+#endif
+
+	return 0;
+}
+
+static int tsl2591_channel_get(const struct device *dev, enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	const struct tsl2591_data *data = dev->data;
+	int64_t cpl = data->atime * data->again;
+	int64_t strength;
+
+	/* Unfortunately, datasheet does not provide a lux conversion formula for this particular
+	 * device. There is still ongoing discussion about the proper formula, though this
+	 * implementation uses a slightly modified version of the Adafruit library formula.
```

Comment:
Please provide a reference

Reply:
Thank you for the review! I've added a reference and fixed the compliance issues.

Though strangely the first time through I originally didn't have the newlines at EOF and checkpatch complained that they SHOULD be there, and then it complained the second time that they SHOULDN'T be there. Third time checkpatch finally seems satisfied on my end, so hopefully all goes well this time through.